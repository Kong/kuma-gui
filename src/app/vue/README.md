---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

