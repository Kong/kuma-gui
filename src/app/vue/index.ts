import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
