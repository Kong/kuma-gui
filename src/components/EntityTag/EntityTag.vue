<template>
  <span class="entity-tag">
    <span
      class="entity-tag__label"
      :class="{ 'entity-tag__label--is-kuma-io-label': isKumaIoLabel }"
    >
      {{ tag.label }}
    </span>

    <span class="entity-tag__value">
      {{ tag.value }}
    </span>
  </span>
</template>

<script lang="ts" setup>
import { computed, PropType } from 'vue'

import { LabelValue } from '@/types'

const props = defineProps({
  tag: {
    type: Object as PropType<LabelValue>,
    required: true,
  },
})

const isKumaIoLabel = computed(() => props.tag.label.toLowerCase().includes('kuma.io/'))
</script>

<style lang="scss" scoped>
$border-radius: 5px;

.entity-tag {
  display: inline-flex;
  align-items: stretch;
  font-size: var(--type-xs);
  border-radius: $border-radius;
}

.entity-tag__label,
.entity-tag__value {
  border: 1px solid transparent;
  padding: 0.1em 0.5em;
}

.entity-tag__label {
  border-right: none;
  border-top-left-radius: $border-radius;
  border-bottom-left-radius: $border-radius;
  color: #fff;
  background-color: var(--blue-1);
}

.entity-tag__label--is-kuma-io-label {
  background-color: var(--brand-color-6);
}

.entity-tag__value {
  border-color: var(--grey-400);
  border-left: none;
  border-top-right-radius: $border-radius;
  border-bottom-right-radius: $border-radius;
  background-color: var(--white);
}
</style>
