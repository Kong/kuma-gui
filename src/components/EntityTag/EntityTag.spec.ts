import { render } from '@testing-library/vue'

import EntityTag from './EntityTag.vue'

describe('EntityTag.vue', () => {
  it('renders basic snapshot', () => {
    const { container } = render(EntityTag, {
      props: {
        tag: {
          label: 'kuma.io/service',
          value: 'backend',
        },
      },
    })

    expect(container).toMatchSnapshot()
  })
})
