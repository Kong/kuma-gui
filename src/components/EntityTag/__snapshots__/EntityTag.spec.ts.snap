// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityTag.vue renders basic snapshot 1`] = `
<div>
  <span
    class="entity-tag"
  >
    <span
      class="entity-tag__label entity-tag__label--is-kuma-io-label"
    >
      kuma.io/service
    </span>
    <span
      class="entity-tag__value"
    >
      backend
    </span>
  </span>
</div>
`;
